Sources/Core/KeychainStore.swift

```diff
@@ -0,0 +1,257 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+// MARK: Imports -
+
+import Foundation
+
+/**
+ The Keychain Store struct, used to query the keychain via Cocoa Framework
+ */
+struct KeychainStore: SecureStore {
+  // MARK: - Enums
+
+  /// Keychain Errors
+  enum KeychainError: FBError {
+    case unexpectedPasswordData
+    case unexpectedItemData
+    case unhandledError(status: OSStatus)
+  }
+
+  /// Keychain Password Types
+  enum PasswordType: String, CaseIterable {
+    case genericPassword
+
+    /// The `CFString` associated value
+    var cfString: CFString {
+      switch self {
+      case .genericPassword:
+        return kSecClassGenericPassword
+      }
+    }
+  }
+
+  /// Keychain Password Types
+  enum PasswordAccessibility: String, CaseIterable {
+    case afterFirstUnlockThisDeviceOnly
+
+    /// The `CFString` associated value
+    var cfString: CFString {
+      switch self {
+      case .afterFirstUnlockThisDeviceOnly:
+        return kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
+      }
+    }
+  }
+
+  // MARK: - Properties
+
+  /// The keychain service
+  let service: String
+
+  /// The optional keychain access group
+  let accessGroup: String?
+
+  // MARK: - Inits
+
+  /**
+   Creates a new KeychainStore
+
+   - Parameter service: The keychain service to use
+   - Parameter accessGroup: The optional keychain accessGroup to use
+   */
+  init(service: String, accessGroup: String? = nil) {
+    self.service = service
+    self.accessGroup = accessGroup
+  }
+
+  // MARK: - Functions
+
+  // MARK: Keychain Access
+
+  /**
+   Retrieves `Data` from the keychain using a provided key
+
+   - Parameter key: The key used to access the value in the keychain
+   - Returns: The data found within the keychain, or `nil` if nothing was found
+   - Throws: `KeychainError`
+   */
+  func keychainData(forKey key: String) throws -> Data? {
+    var query = self.query(forKey: key)
+    query[kSecReturnData as String] = kCFBooleanTrue
+    query[kSecReturnAttributes as String] = kCFBooleanTrue
+    query[kSecMatchLimit as String] = kSecMatchLimitOne
+
+    var queryResult: AnyObject?
+    let status = withUnsafeMutablePointer(to: &queryResult) {
+      SecItemCopyMatching(query as CFDictionary, UnsafeMutablePointer($0))
+    }
+
+    guard status != errSecItemNotFound else {
+      return nil
+    }
+
+    guard status == noErr else {
+      throw KeychainError.unhandledError(status: status)
+    }
+
+    guard let result = queryResult as? [String: AnyObject], let data = result[kSecValueData as String] as? Data else {
```

Comment:
So, for guards I have a better argument for keeping them on multiple lines. If there's ever a failure that we need to debug it's helpful to be able to see which clause the guard is exiting on. If they are on multiple lines it's easier to get this info by stepping through with the debugger. :) Your call.

Reply:
```suggestion
    guard let result = queryResult as? [String: AnyObject], 
      let data = result[kSecValueData as String] as? Data 
      else {
```

Reply:
Will amend